### PR TITLE
planner: ban the prop push down for partition-table with dirty content

### LINKED ISSUE #45007: UnionScan with partition table + keepOrder can't get correct result
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
create table t(a int, b int, primary key(b)) PARTITION BY HASH(b) partitions 4;
analyze table t;
begin;
insert into t(a, b) values (6,6),(3,3),(9,9),(4,4),(5,5),(7,7),(8,8);
select * from t where b > 1 order by b limit 5;
```

### 2. What did you expect to see? (Required)

```
mysql> select * from t where b > 1 order by b limit 5;
+------+---+
| a    | b |
+------+---+
|    3 | 3 |
|    4 | 4 |
|    5 | 5 |
|    6 | 6 |
|    7 | 7 |
+------+---+
5 rows in set (0.00 sec)
```

### 3. What did you see instead (Required)

```
mysql> select * from t where b > 1 order by b limit 5;
+------+---+
| a    | b |
+------+---+
|    4 | 4 |
|    8 | 8 |
|    5 | 5 |
|    9 | 9 |
|    6 | 6 |
+------+---+
5 rows in set (0.00 sec)
```

### PR DESCRIPTION
**Pull request by winoros**
### What problem does this PR solve?

Issue Number: close #45007

Problem Summary:

### What is changed and how it works?

Our read of the dirty table cannot keep order for the partition table.
We ban it first, then enable it after supporting the ordered read.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/exhaust_physical_plans.go [hunk 1]
```diff
@@ -55,6 +55,13 @@ func (p *LogicalUnionScan) exhaustPhysicalPlans(prop *property.PhysicalProperty)
 			"MPP mode may be blocked because operator `UnionScan` is not supported now.")
 		return nil, true, nil
 	}
+	if ds, ok := p.children[0].(*DataSource); ok &&
```
Comment (Defined2014) [inline planner/core/exhaust_physical_plans.go:58]: Better add a comment `TODO` here.

**Review by Defined2014 — APPROVED** [review]

File: planner/core/issuetest/planner_issue_test.go [hunk 2]
```diff
@@ -106,3 +106,16 @@ func TestIssue44051(t *testing.T) {
 	rs := tk.MustQuery("WITH tmp AS (SELECT t2.* FROM t2) SELECT * FROM t1 WHERE t1.id = (select id from tmp where id = 1) or t1.id = (select id from tmp where id = 2) or t1.id = (select id from tmp where id = 3)")
 	rs.Sort().Check(testkit.Rows("1 <nil> <nil> <nil>", "2 <nil> <nil> <nil>", "3 <nil> <nil> <nil>"))
 }
+
+func TestIssue45007(t *testing.T) {
+	store := testkit.CreateMockStore(t)
```
Comment (Defined2014) [inline planner/core/issuetest/planner_issue_test.go:111]: Maybe add `failpoint.Enable("github.com/pingcap/tidb/planner/core/forceDynamicPrune", `return(true)`)` here, I found it not always use dynamic pruning for this situation.

File: planner/core/exhaust_physical_plans.go [hunk 3]
```diff
@@ -55,6 +55,13 @@ func (p *LogicalUnionScan) exhaustPhysicalPlans(prop *property.PhysicalProperty)
...
+		ds.tableInfo.GetPartitionInfo() != nil &&
+		p.ctx.GetSessionVars().StmtCtx.UseDynamicPruneMode &&
+		tableHasDirtyContent(p.ctx, ds.tableInfo) &&
```
Comment (time-and-fate) [inline planner/core/exhaust_physical_plans.go:61]: Is this check necessary? If there's no dirty content, I think there should be no `UnionScan`.

**Comment by Defined2014** [comment]
I think we could close this PR after https://github.com/pingcap/tidb/pull/45140 merged. If we still need this, pls feel free to reopen it.

**Pull request closed without merging**